src/atlas-orca/CMakeLists.txt

```diff
@@ -9,44 +9,87 @@
-        util/PointIJ.h
-        ${CMAKE_CURRENT_BINARY_DIR}/version.cc
-        version.h
-    PUBLIC_LIBS atlas
-    PUBLIC_INCLUDES 
-       $<INSTALL_INTERFACE:include/atlas-orca>
-       $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src>
-       $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/src> )
+if ( HAVE_ORCA_IN_MEMORY_CACHE )
+    ecbuild_add_library(
+        TARGET atlas-orca
+        TYPE SHARED
+        INSTALL_HEADERS LISTED
+        HEADER_DESTINATION "include/atlas-orca"
+        PUBLIC_DEFINITIONS
+            HAVE_ORCA_IN_MEMORY_CACHE
+        SOURCES 
+            Library.cc
+            Library.h
+            grid/Orca.cc
+            grid/Orca.h
+            grid/OrcaGrid.cc
+            grid/OrcaGrid.h
+            grid/OrcaGridBuilder.cc
+            meshgenerator/OrcaMeshGenerator.cc
+            meshgenerator/OrcaMeshGenerator.h
+            util/AtlasIOReader.h
+            util/ComputeCachedPath.h
+            util/ComputeUid.h
+            util/ComputeUid.cc
+            util/DetectInvalidElements.cc
+            util/DetectInvalidElements.h
+            util/Download.cc
+            util/Download.h
+            util/Enums.h
+            util/Flag.h
+            util/OrcaData.cc
+            util/OrcaData.h
+            util/OrcaDataFile.h
+            util/OrcaPeriodicity.cc
+            util/OrcaPeriodicity.h
+            util/PointIJ.h
+            ${CMAKE_CURRENT_BINARY_DIR}/version.cc
+            version.h
+        PUBLIC_LIBS atlas
+        PUBLIC_INCLUDES 
+           $<INSTALL_INTERFACE:include/atlas-orca>
+           $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src>
+           $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/src> )
+else()
+    ecbuild_add_library(
+        TARGET atlas-orca
+        TYPE SHARED
+        INSTALL_HEADERS LISTED
+        HEADER_DESTINATION "include/atlas-orca"
+        SOURCES 
+            Library.cc
+            Library.h
+            grid/Orca.cc
+            grid/Orca.h
+            grid/OrcaGrid.cc
+            grid/OrcaGrid.h
+            grid/OrcaGridBuilder.cc
+            meshgenerator/OrcaMeshGenerator.cc
+            meshgenerator/OrcaMeshGenerator.h
+            util/AtlasIOReader.h
+            util/ComputeCachedPath.h
+            util/ComputeUid.h
+            util/ComputeUid.cc
+            util/DetectInvalidElements.cc
+            util/DetectInvalidElements.h
+            util/Download.cc
+            util/Download.h
+            util/Enums.h
+            util/Flag.h
+            util/OrcaData.cc
+            util/OrcaData.h
+            util/OrcaDataFile.h
+            util/OrcaPeriodicity.cc
+            util/OrcaPeriodicity.h
+            util/PointIJ.h
+            ${CMAKE_CURRENT_BINARY_DIR}/version.cc
+            version.h
+        PUBLIC_LIBS atlas
+        PUBLIC_INCLUDES 
+           $<INSTALL_INTERFACE:include/atlas-orca>
+           $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/src>
+           $<BUILD_INTERFACE:${PROJECT_BINARY_DIR}/src> )
+endif()
```

Comment:
I would avoid the code duplication.
Anyhow, seeing that a runtime option is preferred, this is not needed.
For future reference, you can just use a variable which can be empty:
```
ecbuild_add_library( ...
   PRIVATE_DEFINITIONS ${ATLAS_ORCA_PRIVATE_DEFINITIONS})
```
The definition should not be PUBLIC, but PRIVATE in this case.
You can also add per source file definitions:
```
if ( HAVE_IN_MEMORY_CACHE )
    set_source_files_properties( grid/Orca.cc
                            PROPERTIES COMPILE_DEFINITIONS "ATLAS_ORCA_HAVE_IN_MEMORY_CACHE")
endif()

Reply:
I completely agree that code duplication is undesirable. I wasn't aware that I could have placed the definition in a specific source file.

Reply:
You can look at mir's src/mir/caching/SharedMemoryKey.cc and src/mir/CMakeLists.txt, it's this exact situation too

Reply:
Agree. See my comment below.